cache/lib/cache/authentication.ex

```diff
@@ -0,0 +1,191 @@
+defmodule Cache.Authentication do
+  @moduledoc """
+  Authentication for CAS operations using the /api/projects endpoint.
+
+  This module validates that a request has proper authorization to access a project
+  by calling the server's /api/projects endpoint and caching the results.
+  """
+
+  @failure_cache_ttl 3
+  @success_cache_ttl 600
+  @cache_name :cas_auth_cache
+
+  require Logger
+
+  def child_spec(_) do
+    %{
+      id: __MODULE__,
+      start: {Cachex, :start_link, [@cache_name, []]}
+    }
+  end
+
+  @doc """
+  Ensures the request has access to the specified project.
+
+  Returns `{:ok, auth_header}` if authorized, or `{:error, status, message}` otherwise.
+  """
+  def ensure_project_accessible(conn, account_handle, project_handle) do
+    auth_header = Plug.Conn.get_req_header(conn, "authorization") |> List.first()
+
+    Logger.info("Checking authorization for: #{account_handle}/#{project_handle} with auth header: #{auth_header}")
+
+    if is_nil(auth_header) do
+      {:error, 401, "Missing Authorization header"}
+    else
+      requested_handle = normalize_handle(account_handle, project_handle)
+
+      case authorize(auth_header, requested_handle, conn) do
+        :ok -> {:ok, auth_header}
+        {:error, status, message} -> {:error, status, message}
+      end
+    end
+  end
+
+  defp authorize(auth_header, requested_handle, conn) do
+    cache_key = generate_cache_key(auth_header)
+
+    case cached_failure(cache_key) do
```

Comment:
The data structures here are a bit confusing. If I understood the logic right:
- We have (or not) a cached information that helps us authorize the requests.
- The cached information can be either success or failure. 

Can't we model the business logic as?

```ex
// Returns
// {:error, ...}
// {:ok, [handle_1, handle_2]}
cached_authorization(cache_key) |> fetch_and_cache_if_missing(auth_header)
```